File: EPI/10.4/swift/10.4.swift
public class Node {
    public let left: Node?
    public let right: Node?
    public var parent: Node? = nil

    public init(left: Node? = nil, right: Node? = nil) {
        self.left = left
        self.right = right

        left?.parent = self
        right?.parent = self
    }

    func depth() -> Int {
        var node = self
        var depth = 0
        while let p = node.parent {
            depth += 1
            node = p
        }
        return depth
    }

    func ancestor(depth: Int) -> Node? {
        var node = self
        for _ in 0..<depth {
            if let p = node.parent {
                node = p
            } else {
                return nil
            }
        }
        return node
    }

    public func lca(other: Node) -> Node? {
        let selfDepth = depth()
        let otherDepth = other.depth()

        var selfAncestor = ancestor(max(0, selfDepth - otherDepth))
        var otherAncestor = other.ancestor(max(0, otherDepth - selfDepth))

        while let sa = selfAncestor, oa = otherAncestor {
            if sa === oa {
                return sa
            }
            selfAncestor = sa.parent
            otherAncestor = oa.parent
        }

        return nil
    }
}

File: EPI/10.4/typescript/10.4.ts
export interface Node {
    parent?: Node;
}

export function LCA(a: Node, b: Node): Node | undefined {
    const ad = depth(a);
    const bd = depth(b);

    let [low, high] = ad > bd ? [a, b] : [b, a];

    for (let d = Math.abs(ad - bd); d > 0; d -= 1) {
        low = low.parent!;
    }

    while (low !== high) {
        if (!low.parent || !high.parent) {
            return undefined;
        }

        low = low.parent;
        high = high.parent;
    }

    return low;
}

function depth(node: Node): number {
    let d = 0;

    while (node.parent) {
        node = node.parent;
        d += 1;
    }

    return d;
}
